src/Shared/RoslynUtils/WellKnownTypes.cs

```diff
@@ -86,6 +86,26 @@ private INamedTypeSymbol GetAndCache(int index)
         return _lazyWellKnownTypes[index]!;
     }
 
+    // Look for types within well-known assemblies only.
+    private INamedTypeSymbol? GetTypeByMetadataNameInTargetAssembly(string metadataName)
+    {
+        var types = _compilation.GetTypesByMetadataName(metadataName);
+        if (types.Length == 1)
+        {
+            return types[0];
+        }
+
+        foreach (var type in types)
+        {
+            if (type.ContainingAssembly.Identity.Name.StartsWith("System", StringComparison.Ordinal)
+                || type.ContainingAssembly.Identity.Name.StartsWith("Microsoft", StringComparison.Ordinal))
+            {
+                return type;
```

Comment:
Nit: This doesn't guard against there being two such matching types.

Reply:
I ended up not addressing this here for the sake of keeping the code a little simpler. I think this is a niche enough scenario that the trade-off between code complexity and net value doesn't balance out too well.